### ISSUE TITLE
Add props to control sidebar's collapsed state in DashboardLayout component

### ISSUE THREAD

**Issue by Mohammad-210** [mui/toolpad#4406]
### Summary

This feature request is to enhance the DashboardLayout component by adding two new props:

collapsed: A boolean prop that controls the initial state of the sidebar (collapsed or expanded).
onToggleSidebar: A function prop that allows toggling the sidebar’s collapsed state from outside the DashboardLayout component.
This setup will make the sidebar’s state externally controllable, enabling more dynamic and flexible UI behaviors in various responsive scenarios.

### Examples

import React, { useState } from 'react';
import DashboardLayout from './DashboardLayout';

const AppProvider = () => {
  const [isSidebarCollapsed, setIsSidebarCollapsed] = useState(false);

  const handleToggleSidebar = (collapsed) => {
    setIsSidebarCollapsed(collapsed);
  };

  return (
    <DashboardLayout
      collapsed={isSidebarCollapsed}
      onToggleSidebar={handleToggleSidebar}
    >
      {/* Application Content */}
    </DashboardLayout>
  );
};

export default AppProvider;

### Motivation

This feature will improve DashboardLayout’s flexibility and control by allowing parent components to manage and toggle the sidebar’s collapsed state. With responsive design requirements, many apps need to control the sidebar's visibility dynamically based on user interactions or screen size. This enhancement will make DashboardLayout more adaptable for complex UI requirements and enhance user experience across devices. The issue i face is i need a state to ensure the sidebar is collapsed or not to manage the responsiveness of design of rendered children comp  

**Search keywords**: sidebar toggle, sidebar collapsed state, DashboardLayout, Material-UI, Toolpad sidebar, responsive sidebar, toggle sidebar from parent, sidebar props

**Comment by bharatkashyap** [comment]
Agree with providing a way to access the state of the sidebar, and also imperatively setting it. An alternative way of doing this is through a `useSidebar` hook:

```tsx
const { collapsed, setCollapsed } = useSidebar();

const onClick = () => { if(!collapsed) { setCollapsed(); } 

```

This will enable components across the app to be able to access (and imperatively set) the sidebar's state instead of having to share props.

**Comment by vikasgurjar** [comment]
@Mohammad-210 are you planning to work on this create a PR? If not, I can help with the PR

**Comment by Janpot** [comment]
a controlled prop makes sense to me. a hook will be constraint to descendants only, which is probably too limiting.

**Comment by Mohammad-210** [comment]
@vikasgurjar I'm swamped with other tasks, I won't get to it.But I'm planning to fix it, but I haven't started yet. Feel free to create the PR!. if help needed let me know ..!

**Comment by vikasgurjar** [comment]
@Janpot from what I understand, controlled prop will have to be passed down to every child that needs it. Also even in that case these props will be constraint to descendants only

**Comment by vikasgurjar** [comment]
I'm thinking of using `useNavigation` instead of `useSidebar` for consistency as we already have a prop called 'hideNavigation'

> ```tsx
> const { collapsed, setCollapsed } = hideNavigation();
> 
> const onClick = () => { if(!collapsed) { setCollapsed(); } 
> ```

**Comment by bharatkashyap** [comment]
> @bharatkashyap I'm thinking of using `useNavigation` instead of `useSidebar` for consistency as we already have a prop called 'hideNavigation'
> 
> > ```tsx
> > const { collapsed, setCollapsed } = hideNavigation();
> > 
> > const onClick = () => { if(!collapsed) { setCollapsed(); } 
> > ```

I discussed this with @mnajdova and we came to the conclusion that a controlled prop makes sense to begin with, while a hook/context can be implemented by users themselves if they need it as long as the prop is present. We could make an issue to collect upvotes on the hook feature request, to see if there is enough interest in providing that out of the box. 

Let's focus on adding a controlled prop as part of this issue 👍🏻

**Comment by apedroferreira** [comment]
A controlled prop as suggested sounds good!
`sidebarCollapsed` and `onToggleSidebar` sound good as possible prop names for `DashboardLayout`.
Feel free to create a PR, or we can get to it eventually.

**Comment by vikasgurjar** [comment]
@apedroferreira I think eventually `sidebarCollapsed` can replace `defaultSidebarCollapsed` in future. Do you think we will deprecate the prop? If so, should we add the warning for the same?

**Comment by apedroferreira** [comment]
> @apedroferreira I think eventually `sidebarCollapsed` can replace `defaultSidebarCollapsed` in future. Do you think we will deprecate the prop? If so, should we add the warning for the same?

I think they would both stay with different purposes: `sidebarCollapsed` if you want that state the be controlled, and `defaultSidebarCollapsed` just to set the initial value if you want to leave it uncontrolled. 
`sidebarCollapsed` would always take priority if set.

**Comment by vikasgurjar** [comment]
@apedroferreira @bharatkashyap I got some time to work on this. I was wondering what's the use of `onToggleSidebar` prop? Will it act as just a callback for whenever there is a state change for sidebarCollapsed state?

**Comment by apedroferreira** [comment]
> @apedroferreira @bharatkashyap I got some time to work on this. I was wondering what's the use of `onToggleSidebar` prop? Will it act as just a callback for whenever there is a state change for sidebarCollapsed state?

Yes, it should run when the menu button on the top-right is clicked and by default toggle the sidebarCollapsed state.

**Comment by apedroferreira** [comment]
> Any progress?

Not being prioritized by us right now.
I've reviewed https://github.com/mui/toolpad/pull/4516 some time ago but the PR seems to have been discontinued.

**Comment by mortoni-tmr** [comment]
Hi team, this proposed enhancement would be extremely helpful for our use case. We currently need external control over the sidebar's collapsed state to better handle responsive layouts and user-driven toggling from parent components.

Having collapsed and onToggleSidebar as props in DashboardLayout would simplify our logic and improve flexibility significantly.

Looking forward to seeing this implemented — happy to help test or review if needed!

Thanks 🙌
